### ISSUE TITLE
how to implement scaffold hopping/elaboration and fragment growing/merging Case?

### ISSUE THREAD

**Issue by berlinliumingfei** [arneschneuing/DiffSBDD#33]

**Comment by berlinliumingfei** [comment]
@arneschneuing

**Comment by akichinguyen** [comment]
Same question

**Comment by berlinliumingfei** [comment]
@yuanqidu 
Dear Yuanqidu,

I hope this email finds you well. I am writing to express my sincere appreciation for your exceptional work. Your contributions have truly showcased remarkable creativity and thoughtfulness. I have been studying your projects closely and find them both intriguing and inspiring.

Your ability to think outside the box and bring fresh perspectives to your work is truly commendable. It is evident that you possess a unique talent for crafting innovative solutions. Your dedication to pushing boundaries and exploring new possibilities is something I greatly admire.

As I continue to delve deeper into your work, I am motivated to apply your ideas to real-world scenarios. However, I have encountered a minor roadblock and was wondering if you could spare some time to assist me in finding a solution. I believe your expertise and insights would be invaluable in overcoming this challenge.

Would it be possible to schedule a brief meeting or have a quick discussion over email? I would be grateful for any guidance or advice you can offer. Your assistance would not only help me progress with my project but also provide me with an opportunity to learn from a visionary like yourself.

Once again, thank you for your outstanding work and the inspiration you have provided. I look forward to hearing from you soon.

Warm regards,

Chao Liu

**Comment by yuanqidu** [comment]
Dear Chao, 

Thanks so much for your kind words! I am happy to help regarding your request. My first impression is that it is similar to inpaint the ligand, but the difference is that you need to input your fragment for merging or growing. I believe @cch1999 is the better expert on this as he wrote the code in particular for that part. Let @cch1999 see if he has a brief answer or could help.

Best,
Yuanqi

**Comment by berlinliumingfei** [comment]
Dear Charlie Harris,

Can you show us an example about how to implement scaffold hopping/elaboration and fragment growing/merging.   Looking forward to your reply！ @cch1999 

Best，
Chao

**Comment by berlinliumingfei** [comment]
@akichinguyen Thank you for sharing. I have a preliminary understanding of how to make the modification, but I cannot find the "ligand" variable. Could you provide me with a complete example?

Here are all the variables:
if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument('checkpoint', type=Path)
    parser.add_argument('--pdbfile', type=str)
    parser.add_argument('--resi_list', type=str, nargs='+', default=None)
    parser.add_argument('--ref_ligand', type=str, default=None)
    parser.add_argument('--outdir', type=Path)
    parser.add_argument('--n_samples', type=int, default=20)
    parser.add_argument('--num_nodes_lig', type=int, default=None)
    parser.add_argument('--all_frags', action='store_true')
    parser.add_argument('--sanitize', action='store_true')
    parser.add_argument('--relax', action='store_true')
    parser.add_argument('--resamplings', type=int, default=1)
    parser.add_argument('--jump_length', type=int, default=1)
    parser.add_argument('--timesteps', type=int, default=None)
    args = parser.parse_args()

**Comment by akichinguyen** [comment]
@berlinliumingfei All the modifications I mentioned are in "lightning_modules.py". 
This one:
ligand = {
                'x': torch.zeros((len(lig_mask), self.x_dims),
                                 device=self.device, dtype=FLOAT_TYPE),
                'one_hot': torch.zeros((len(lig_mask), self.atom_nf),
                                       device=self.device, dtype=FLOAT_TYPE),
                'size': num_nodes_lig,
                'mask': lig_mask
            }
and this one:
lig_mask_fixed = torch.zeros(len(lig_mask), device=self.device)
Here you can modify your value as I described above instead of all zeros.

**Comment by berlinliumingfei** [comment]
@akichinguyen How can I assign a value to the ”lig_mask“ variable? In the code, there is this line:  **parser.add_argument('--num_nodes_lig', type=int, default=None)**,  where” --num_nodes_lig“ is of type int. How can I pass in the sdf file of the ligand fragment as an argument?

**Comment by berlinliumingfei** [comment]
@akichinguyen I am slowly understanding your idea. I should modify the x, one_hot, and lig_mask_fixed to real values, but I still have a small question. How can I get the real x, one_hot, and lig_mask_fixed for the fragment? Thank you very much for your help. I am not very familiar with this work and may have some misunderstandings. Please excuse me.

**Comment by akichinguyen** [comment]
The num_nodes_lig is just the number of atoms you want for your ligand. It can be whatever > number of atoms of your fragment. You can parse something like this parser.add_argument('--anchor', type=str) which points to the fragment file, and modify the  "generate_ligands" function (inside the lightning_modules.py) to take it as an argument and process the fragment file. From what I tested, I do not think any of their arguments in generate_ligands.py could help to achieve fragment growth without modifying the core function.

**Comment by akichinguyen** [comment]
The rdkit.Chem library could help to read sdf files and get positions and atom types. Hope it could help you.

**Comment by berlinliumingfei** [comment]
@akichinguyen I agree with your perspective. Meanwhile, when testing the scenario of pocket-conditioned small molecule generation, I found that the generated molecules had poor drug-likeness. So, I called in my colleagues who are chemists to analyze it together. They have also consistently believed that there is an issue with the drug-likeness of the generated molecules. The model has not been able to learn the spatial distribution of molecules well, especially in the distribution of chemical bonds!
